### DISCUSSION TITLE
Kratos can't be configured

### DISCUSSION THREAD

**Discussion by ghost** [ory/kratos#1344] [Q&A] [answered]
Hi, I am currently trying to run kratos on my machine with the "kratos serve" command. If I understand correctly I have to give kratos my config file when I try that.
So I created a file named "kratos.yml" as recommended and used the exact same config as in the file "kratos.yml" on "https://www.schemastore.org/json/". Right? When I try to run kratos now I get the following error message:

`The configuration contains values or keys which are invalid:
The configuration contains values or keys which are invalid:
identity: <nil>
          ^-- one or more required properties are missing

The configuration contains values or keys which are invalid:
courier.smtp.connection_uri: <nil>
                             ^-- one or more required properties are missing

The configuration contains values or keys which are invalid:
selfservice.default_browser_return_url: <nil>
                                        ^-- one or more required properties are missing

The configuration contains values or keys which are invalid:
(root)
^-- additionalProperties "title", "type", "oneOf", "$id", "$schema" not allowed

FATA[2021-05-16T10:35:22+02:00] Unable to instantiate configuration.          audience=application error=map[message:I[#] S[#] validation failed
  I[#] S[#/required] missing properties: "identity", "dsn"
  I[#/courier/smtp] S[#/properties/courier/properties/smtp/required] missing properties: "connection_uri"
  I[#/selfservice] S[#/properties/selfservice/required] missing properties: "default_browser_return_url"
  I[#] S[#/additionalProperties] additionalProperties "title", "type", "oneOf", "$id", "$schema" not allowed] service_name=Ory Kratos service_version=v0.6.2-alpha.1`

So basically I seem to do something completely wrong here... Can somebody help me please?
Thanks.

**Comment by Benehiko** [comment, marked as answer]
Hi @LeonRickert 

I think the best would be to start with the [Quickstart guide](https://www.ory.sh/kratos/docs/quickstart) as this gives a working kratos configuration file. Then you can move to adding or removing some of the configs given by the quickstart.

**Reply by vinckr** [reply]
Hey @LeonRickert, just checking in if your problem still persists. 
I will mark this as answered for now, but feel free to come back to the discussion any time.
Thanks a ton :bee:

**Reply by ghost** [reply]
Running these commands from the quickstart guide results in Kratos running successfully on the node.js UI.

```
git clone https://github.com/ory/kratos.git
cd kratos
git checkout v0.7.0-alpha.1

docker-compose -f quickstart.yml -f quickstart-standalone.yml up --build --force-recreate
```

I'm attempting to do something different than running the quickstart.yml.
What I am attempting to do is:

1. Build Kratos into a docker image
2. Push that image to an Image hosting repository like DockerHub or ECR
3. Deploy that image with Digital Ocean or ECS
4. Hook that image up to a PostgreSQL database
5. Use the deployed site as an authentication API to implement authentication into a custom UI.

The quickstart identity is not a valid identity location as ` default_schema_url: file:///etc/config/kratos/identity.schema.json` doesn't provide an actual location for an identity schema..

Now where I'm going wrong is the image itself.  It doesn't run locally with `docker run kratos:latest`.  I'm getting the same error by following these steps.

```
git clone https://github.com/ory/kratos.git
cd kratos
git checkout v0.7.0-alpha.1
cd .docker;cat Dockerfile-build | pbcopy;nano ../Dockerfile; pbpaste; #if pbpaste doesnt work use ctrl + v
docker build -t kratos .
docker run kratos:latest
```

I get 

```
The configuration contains values or keys which are invalid:
The configuration contains values or keys which are invalid:
identity: <nil>
          ^-- one or more required properties are missing

The configuration contains values or keys which are invalid:
selfservice.default_browser_return_url: <nil>
                                        ^-- one or more required properties are missing

The configuration contains values or keys which are invalid:
courier.smtp.connection_uri: <nil>
                             ^-- one or more required properties are missing

time=2021-07-21T20:18:50Z level=fatal msg=Unable to instantiate configuration. audience=application error=map[message:I[#] S[#] validation failed
  I[#] S[#/required] missing properties: "identity", "dsn"
  I[#/selfservice] S[#/properties/selfservice/required] missing properties: "default_browser_return_url"
  I[#/courier/smtp] S[#/properties/courier/properties/smtp/required] missing properties: "connection_uri"] service_name=Ory Kratos service_version=master
```

The issue is a configuration issue, its not getting the values that it needs...they're nil.

Documentation states that 

`If file $HOME/.kratos.yaml exists, it will be used as a configuration file which supports all configuration settings listed below.` 

but I have a kratos.yaml file at ~/.kratos.yaml (exact same contents) and it the run seems unaffected by it.

I've also tried putting the value into an environment variable with: `export IDENTITY_DEFAULT_SCHEMA_URL="full-path-to-schema/identity.schema.json"` to no avail.

My question is:

1.  What config file is Kratos expecting the identity to come from?

**Reply by Ryu36** [reply]
I have the same issue

**Reply by vinckr** [reply]
Hey, thanks for reporting this! I will look into it, it might be a bug.

**Reply by aeneasr** [reply]
No, this is not a bug, please read the answer in the other thread you opened: https://github.com/ory/kratos/discussions/1600#discussioncomment-1081474

**Comment by vinckr** [comment]
The schema you tried is this one: 
https://www.ory.sh/kratos/docs/reference/configuration/

`This reference configuration documents all keys, also deprecated ones! It is a reference for all possible configuration values.`

As Benehiko correctly pointed out the Quickstart is the correct place for a sample configuration.
